### PR TITLE
Tempus: Remove ParameterList from IntegratorBasic

### LINKED ISSUE #8704: Tempus: Remove ParameterList from Internals of IntegratorBasic
## Enhancement
@trilinos/tempus 

Remove all the internal uses of ParameterList from IntegratorBasic. This means moving the variables in the IntegratorBasic ParameterList to member data. Integrator will not longer inherit from Teuchos::ParameterListAcceptor. However IntegratorBasic can still be built from a ParameterList, and will still provide a valid ParameterList.

This is in support of improving the data flow transparency for the Tempus PLATR deliverables.

### PR DESCRIPTION
**Pull request by ccober6**
Remove all the internal uses of ParameterList from IntegratorBasic.
This means moving the variables in the IntegratorBasic ParameterList
to member data. Integrator will not longer inherit from
Teuchos::ParameterListAcceptor. However IntegratorBasic can still
be built from a ParameterList, and will still provide a valid
ParameterList.

 * To break up these changes, created a copy of IntegratorBasic
   (i.e., IntegratorBasicOld) for the sensitivity analysis integrators
    - IntegratorAdjointSensitivity
    - IntegratorForwardSensitivity
    - IntegratorPseudoTransientAdjointSensitivity
    - IntegratorPseudoTransientForwardSensitivity
   so these can be upgraded in another PR.
 * IntegratorBasic is no longer inherited from ParameterListAcceptor.
    - Removed setParameterList.
    - IntegratorBasic constructors using ParameterLists have moved to
      nonmember constructors, e.g.,
      . integratorBasic(pl, model) --> createIntegratorBasic(pl, model)
      . IntegratorBasic(pl, model) --> createIntegratorBasic(pl, model)
    - Member data ParameterLists are removed.
    - Kept getValidParameters(), which now returns a ParameterList
      with the current values. Still matches ParameterListAcceptor
      signature.
 * Ensured that ParameterList names were correctly set, so
   getValidParameters() could be used to create nested ParameterLists,
   e.g., IntegratorBasic->Stepper->Solver.
 * Made getValidParametersBasic() a member functions of Stepper class.
 * Simplified setting the Stepper to just setStepper(stepper).
 * Added method to set model on stepper in IntegratorBasic,
   i.e., setModel(model).
 * The integrator observer is no longer a composite observer.
   It is simply a base class observer.
 * All internal IntegratorBasic references to member ParameterLists
   are changed to member data.
 * Added member data for the integrator name and type.  Name is a
   label that used for identification, e.g., 'My Integrator Basic'.
   Type defines the derived class being used, e.g., 'Integrator Basic'.
 * Added a shallow copy for the SolutionHistory.

@trilinos/tempus 

## Motivation
Part of improving data transparency requested from ASC applications.

* Closes #8704

### CODE REVIEW

File: packages/tempus/src/Tempus_StepperBDF2_impl.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_StepperBDF2_impl.hpp:24]: We need to document the distinction between "`StepperName` and `StepperType`.

File: packages/tempus/src/Tempus_StepperOperatorSplit_impl.hpp
Comment (sconde) [inline packages/tempus/src/Tempus_StepperOperatorSplit_impl.hpp:479]: shadow warning here

**Comment by sconde** [comment]
Piro needs to be updated

File: packages/tempus/src/Tempus_StepperBDF2_impl.hpp
Comment (ccober6) [reply packages/tempus/src/Tempus_StepperBDF2_impl.hpp:24]: With latest commit, added documentation on stepperName and stepperType.

File: packages/tempus/src/Tempus_StepperOperatorSplit_impl.hpp
Comment (ccober6) [reply packages/tempus/src/Tempus_StepperOperatorSplit_impl.hpp]: Fixed it!  Thanks!

**Comment by ccober6** [comment]
> Piro needs to be updated

Fixed this and ROL also.  Everything should pass now.

**Comment by ccober6** [comment]
@sconde  So everything now passes.  Is there anything else I should look at?

**Comment by ccober6** [comment]
@sconde  I will go ahead and squash and rebase, if you are ready to approve.

**Comment by sconde** [comment]
I'm rebuilding with Drekar now

**Comment by ccober6** [comment]
Great!

**Comment by sconde** [comment]
Ok, Drekar is building and tests are passing. I'm not testing with SPARC

**Comment by ccober6** [comment]
@sconde OK great!  I will ping Mike and Irina and ask if they can test this.

**Comment by ccober6** [comment]
@michael-a-hansen @ikalash Can you test this PR against SPARC and Albany?  

This is a fairly big change with some interface mods.  Let me know if I can help with anything.  Thanks, Curt

**Comment by sconde** [comment]
I've got a patch for SPARC compiling and tests are passing. But I'm not happy with it.

**Comment by ccober6** [comment]
@sconde Is there something we can improve on the Tempus side?

**Comment by ccober6** [comment]
@ikalash Have you tested this against Albany?

**Comment by ikalash** [comment]
No not yet...  will test now.  Sorry about the delay.

**Comment by ccober6** [comment]
No problem.  Thanks for the help!

**Comment by ikalash** [comment]
@ccober6 : so there is one test failure in Albany, a GERK test.  It looks like the wrong solution is computed, causing the test to diff: 
```
 Response 0: Solution Average
2: 
2:                     1.728355723827e+00
2: Response Test 0: 1.728355723827e+00 != 1.001082996554e+00 (rel 1.000000000000e-03 abs 1.000000000000e-05)
2: 
```
Here is the input file for the test: https://github.com/SNLComputation/Albany/blob/master/tests/small/TransientHeat2D/tempus_gerk.yaml .  Since you are more familiar with the details of this PR, is there something that would need to be changed in this input file for the test to run correctly with the changes to Tempus?

**Comment by ccober6** [comment]
@ikalash Interesting... With a quick look, I do not see anything.

One thing that got my eye was on line 75 of https://github.com/SNLComputation/Albany/blob/master/tests/small/TransientHeat2D/tempus_gerk.yaml. There seems be to the following two characters "|-" is that some special yaml?

**Comment by ccober6** [comment]
@ikalash Can you add the following line after you finished construction of the stepper?
```
  stepper->describe( *(Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout))),
                      Teuchos::VERB_EXTREME);
```
It looks like you are building with the default tableau (RK4).  You should get some input similar to
```
--- Stepper ---
  isInitialized_      = true
  stepperType_        = General ERK
  useFSAL_            = false
  ICConsistency_      = None
  ICConsistencyCheck_ = false
  stepperX_           = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x0,node=0,strong_count=0,weak_count=0}
  stepperXDot_        = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x0,node=0,strong_count=0,weak_count=0}
  stepperXDotDot_     = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x0,node=0,strong_count=0,weak_count=0}
--- StepperExplicit ---
  appModel_         = Teuchos::RCP<Thyra::ModelEvaluator<double> const>{ptr=0x7fef78837e00,node=0x7fef77829cf0,strong_count=2,weak_count=0}
  inArgs_           = Thyra::ModelEvaluatorBase::InArgs<double>:
  model = Tempus_Test::SinCosModel<double>
  Np = 0
  x_dot = Thyra::DefaultSpmdVector<double>{spmdSpace=Thyra::DefaultSpmdVectorSpace<double>{globalDim=2,localSubDim=2,localOffset=0,comm=NULL}}
  ||x_dot|| = 1
  x = Thyra::DefaultSpmdVector<double>{spmdSpace=Thyra::DefaultSpmdVectorSpace<double>{globalDim=2,localSubDim=2,localOffset=0,comm=NULL}}
  ||x|| = 1
  t = 0
  alpha = 0
  beta = 0

  outArgs_          = Thyra::ModelEvaluatorBase::OutArgs<double>:
  model = Tempus_Test::SinCosModel<double>
  Np = 0
  Ng = 0

--- StepperExplicitRK ---
General ERK
number of Stages = 4
A = 
Values_copied : yes
Rows : 4
Columns : 4
LDA : 4
0 0 0 0 
0.5 0 0 0 
0 0.5 0 0 
0 0 1 0 

b = 
Values_copied : yes
Length : 4
0.166667 0.333333 0.333333 0.166667 

c = 
Values_copied : yes
Length : 4
0 0.5 0.5 1 

bstar = 
Values_copied : no
Length : 0
(std::vector is empty, no values to display)

order    = 4
orderMin = 4
orderMax = 4
isImplicit = 0
isDIRK     = 0
isEmbedded = 0
  tableau_           = Teuchos::RCP<Tempus::RKButcherTableau<double> >{ptr=0x7fef7782a1c0,node=0x7fef7782a340,strong_count=1,weak_count=0}
  stepperRKAppAction_= Teuchos::RCP<Tempus::StepperRKAppAction<double> >{ptr=0x7fef77829940,node=0x7fef77829c40,strong_count=1,weak_count=0}
  stageXDot_.size()  = 4
    stageXDot_[0] = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x7fef7782a000,node=0x7fef77829e80,strong_count=1,weak_count=0}
    stageXDot_[1] = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x7fef7782a640,node=0x7fef7782a550,strong_count=1,weak_count=0}
    stageXDot_[2] = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x7fef7782aa40,node=0x7fef7782a950,strong_count=1,weak_count=0}
    stageXDot_[3] = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x7fef7782ae20,node=0x7fef7782ad30,strong_count=1,weak_count=0}
  useEmbedded_       = false
  ee_                = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x0,node=0,strong_count=0,weak_count=0}
  abs_u0             = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x0,node=0,strong_count=0,weak_count=0}
  abs_u              = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x0,node=0,strong_count=0,weak_count=0}
  sc                 = Teuchos::RCP<Thyra::VectorBase<double> >{ptr=0x0,node=0,strong_count=0,weak_count=0}
-------------------------
```
Just want to make sure the construction went OK.  :)

**Comment by ikalash** [comment]
@ccober6 : no that's not it - that would give rise to a syntax error if it were one, which it's not with yaml.  I think what we're seeing is a bit disturbing b/c somehow the solution changed now for the GERK scheme for this test.  I haven't had a chance to look in detail at what went into this PR.  I'm guessing the GERK schemes were not touched.  Could it be that some PL options that used to be default changed, which is causing the change in behavior?

**Comment by ikalash** [comment]
I should add, we don't particularly care about the GERK scheme right now in Albany so I don't think this would really affect anything for us; but I would be curious what in Tempus made the test change its behavior.

**Comment by sconde** [comment]
@ikalash can we see the output at each time step?

**Comment by ikalash** [comment]
@sconde : yes - I'm attaching the output from before w/o this PR (ctest_before.txt) and with this PR (ctest_after.txt).  The input file for the problem can be found here: https://github.com/SNLComputation/Albany/blob/master/tests/small/TransientHeat2D/tempus_gerk.yaml .  I'm a bit lost with this because I'm not too familiar with this test in Albany, and to try to figure out the diffs I'd have to: (1) refresh my memory on how GERK works, and (2) familiarize myself with the Tempus implementation of it / changes to figure out what could be different.  Let me know if you see anything based on the output.

[ctest_before.txt](https://github.com/trilinos/Trilinos/files/6233217/ctest_before.txt)
[ctest_after.txt](https://github.com/trilinos/Trilinos/files/6233218/ctest_after.txt)

**Comment by ccober6** [comment]
@ikalash So Response[0] is different from the very first time step, so I would guess that the construction of the stepper is different somehow (should look at stepper->describe() from before and after), or the initial conditions are different?  I would guess the former.

I am assuming you are constructing everything (e.g., Stepper, TimeStepControl) through the IntegratorBasic by passing it the ParameterList and ModelEvaluator?

Let me know if you have some time to debug this.

**Comment by ikalash** [comment]
@ccober6 : thanks for looking at it.  I noticed the discrepancy too.  I'll try to find time to print some stuff to triage this, as you suggest - it may not be until this weekend or early next week, as I have some CIS ERB commitments this week, and an LOI for a proposal due on Friday.  If you think it would be helpful for the two of us to look at it together, I can set up a meeting.  I'm OK trying to figure it out on my own first once I can find the time.  I apologize for delaying your PR.  If you really want it merged in, we could live with 1 failing Albany test due to this while we're triaging/debugging the problem.

**Comment by ccober6** [comment]
@ikalash Also can you explicitly set a few things, e.g., "Use FSAL".  This might quickly id the problem.

```
      Tempus Stepper: 
        Stepper Type: General ERK
        Use FSAL: false
        Initial Condition Consistency: None
        Initial Condition Consistency Check: false
        Tableau: 
          A: |-
            0.000000000000000  0.000000000000000  0.000000000000000 0.000000000000000;
                                 0.500000000000000  0.000000000000000  0.000000000000000 0.000000000000000;
                                 0.000000000000000  0.500000000000000  0.000000000000000 0.000000000000000;
                                 0.000000000000000  0.000000000000000  1.000000000000000 0.000000000000000
          b: 0.166666666666667  0.333333333333333  0.333333333333333 0.166666666666667
          c: 0.000000000000000  0.500000000000000  0.500000000000000 1.000000000000000
          order: 4
```

**Comment by ccober6** [comment]
@ikalash No problem.  I am willing to help debugging this.  This has sat long enough I have to rebase again anyways.  It can wait a little longer. :)

**Comment by ikalash** [comment]
> @ikalash Also can you explicitly set a few things, e.g., "Use FSAL". This might quickly id the problem.
> 
> ```
>       Tempus Stepper: 
>         Stepper Type: General ERK
>         Use FSAL: false
>         Initial Condition Consistency: None
>         Initial Condition Consistency Check: false
>         Tableau: 
>           A: |-
>             0.000000000000000  0.000000000000000  0.000000000000000 0.000000000000000;
>                                  0.500000000000000  0.000000000000000  0.000000000000000 0.000000000000000;
>                                  0.000000000000000  0.500000000000000  0.000000000000000 0.000000000000000;
>                                  0.000000000000000  0.000000000000000  1.000000000000000 0.000000000000000
>           b: 0.166666666666667  0.333333333333333  0.333333333333333 0.166666666666667
>           c: 0.000000000000000  0.500000000000000  0.500000000000000 1.000000000000000
>           order: 4
> ```

I tried this and it appears there is no difference.  I am attaching the new input and output files.  I'll try to dig into this after Friday.
[ctest_usefsal_false.txt](https://github.com/trilinos/Trilinos/files/6241035/ctest_usefsal_false.txt)
[tempus_gerk.yaml.txt](https://github.com/trilinos/Trilinos/files/6241040/tempus_gerk.yaml.txt)

**Comment by ccober6** [comment]
@ikalash when you get back to this, also run the 'before' with the defaults set explicitly.  It is possible that the defaults may not have been 'right' in the before case.

**Comment by ikalash** [comment]
@ccober6 : which defaults would this be?

**Comment by ccober6** [comment]
@ikalash Maybe I misunderstood what you have run; maybe you have already run it.  I was just suggesting running with the explicit defaults in your yaml with the before executable, e.g., 
 ```
       Use FSAL: false
        Initial Condition Consistency: None
        Initial Condition Consistency Check: false
```

**Comment by ikalash** [comment]
Oh yes - I did this, same behavior (see attached input and output file below).  I also tried with Use FSAL: true (not sure if it makes sense) and saw the same behavior.

**Comment by ccober6** [comment]
Ok, it appears that there is something more insidious going on here.  :(

Let me know if you have questions, or if I can help track this down.

**Comment by ccober6** [comment]
Also RK4 (the default for General ERK) does not meet the criteria for FSAL, so it should always be false for it.  But in the spirit of tracking down this issue, it is good to double check.  :)

**Comment by ccober6** [comment]
@ikalash I wanted to touch base with you and see if you are going to have chance to look at this.  Should I go ahead and merge?

**Comment by ccober6** [comment]
@ikalash Should I go ahead and merge this, and we can deal with the Albany separately?  Thanks, Curt

**Comment by sconde** [comment]
We need a `createIntegratorBasic` that takes only a PL.

Passing in `model=Teuchos::null` raises a SEGFault with the line `auto newState = createSolutionStateME(integrator->getStepper()->getModel(),integrator->getStepper()->getDefaultStepperState())`

**Comment by ccober6** [comment]
@sconde Sounds good I will add the ability to set `model=Teuchos::null` in the argument list.

**Comment by ccober6** [comment]
@ikalash I am going to make the changes for @sconde suggestion and merge this unless you would like us to hold off.  I think we can track down your issue after the fact.

**Comment by ikalash** [comment]
@ccober6 : yes this is fine.  I'm so sorry I haven't had a chance to look at this still.  I am up to my eyeballs in proposal and am spending most of my time on them, including evenings and weekends... :(  Looking at this is still on my to-do list though.  But please proceed w/o waiting for me - I agree that the issue can be fixed later.

**Comment by ccober6** [comment]
@ikalash Yeah, I figured you were swamped.  Let me know if you need a hand when you get to this.  Thanks!

**Comment by ikalash** [comment]
I feel so bad to be holding this up.  Would it help to send you build instructions for Albany and a pointer to the problem that has the issue?  That should be easy to do on any of the platforms across Sandia.

**Comment by ccober6** [comment]
@sconde Let's go ahead and merge this.  I have the changes implemented for createIntegratorBasic(ParameterList) and will make a separate issue and PR for it.

Can you re-approve this, and we can get it merged?  Thanks!

**Review by sconde — APPROVED** [review]

**Pull request merged**
